iocore/net/quic/QUICFlowController.cc

```diff
@@ -130,11 +151,40 @@ QUICLocalFlowController::forward_limit(QUICOffset offset)
   QUICFlowController::forward_limit(offset);
 
   // Send MAX_(STREAM_)DATA frame
-  if (this->_limit - this->_offset <= this->_threshold) {
+  if (this->_need_to_gen_frame()) {
     this->_frame = this->_create_frame();
   }
 }
 
+int
+QUICLocalFlowController::update(QUICOffset offset)
+{
+  int ret = QUICFlowController::update(offset);
+  if (!ret) {
+    // Send MAX_(STREAM_)DATA frame
+    this->_analyzer.update(offset);
+    if (this->_need_to_gen_frame()) {
+      this->_frame = this->_create_frame();
+    }
+  }
+
+  return ret;
+}
+
+bool
+QUICLocalFlowController::_need_to_gen_frame()
+{
+  if (this->_offset + this->_threshold > this->_limit) {
+    return true;
+  }
+
+  if (this->_analyzer.calculate_time_to_limit(this->_limit - this->_threshold - this->_offset) <= 2 * this->smoothed_rtt()) {
```

Comment:
How about updating `_threshold`?

If you calculate the amount of data that will be arrive within next 2 roundtrip time and put it into `_threshold`, the condition, `this->_offset + this->_threshold > this->_limit`, would work I think.